p2p/p2p/src/inbound_server.rs

```diff
@@ -102,8 +114,38 @@ where
                 .instrument(Span::current()),
             );
         } else {
+            // Otherwise check if the node is simply pinging us.
             tracing::debug!("No permit free for incoming connection.");
-            // TODO: listen for if the peer is just trying to ping us to see if we are reachable.
+            
+            // We only handle 2 ping request conccurently. Otherwise we drop the connection immediately.
+            if let Ok(_permit) = ping_semaphore.clone().try_acquire_owned() {
+                let task = tokio::spawn(
+                    async move {
+                        // Await first message from node. If it is a ping request we respond back, otherwise we drop the connection.
+                        let fut = timeout(Duration::from_secs(5), peer_stream.next()); // S
+                        
+                        #[allow(clippy::collapsible_match)]
+                        // Ok if timeout did not elapsed -> Some if there is a message -> Ok if is has been decoded
+                        if let Ok(Some(Result::<Message,BucketError>::Ok(command))) = fut.await {
+                            
+                            if let Message::Request(AdminRequestMessage::Ping) = command {
+                                let _ = peer_sink
+                                    .send(Message::Response(AdminResponseMessage::Ping(
+                                        PingResponse {
+                                            status: PING_OK_RESPONSE_STATUS_TEXT, 
+                                            peer_id: our_peer_id
+                                        }
+                                    )).into())
+                                    .await
+                                    .inspect_err(|err: &BucketError| tracing::debug!("Unable to respond to ping request from peer ({addr}): {err}"));
+                            }
+                        }
+                    }
+                    .instrument(Span::current())
+                );
+                
+                task.await.expect("A tokio task responsible for handling ping requests panicked!");
```

Comment:
By awaiting the task here we block the inbound server from processing other connections.

I would recommend spawning these tasks on a [JoinSet](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#) and checking the length to limit concurrent pings instead of using a semaphore.

Reply:
Fixed in latest commit.